### PR TITLE
Add support for draining LB Nodes

### PR DESCRIPTION
**Pull request by TuxPowered42**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -31,9 +31,10 @@ HTTP and HTTPS
 
 The check sends the following HTTP request::
 
-	$hc_query HTTP/1.1
-	Host: $ip_addr
-	Connection: close
+  $hc_query HTTP/1.1
+  Host: $ip_addr
+  User-agent: testtool
```
Comment (Woellchen) [inline README.rst:36]: lets add the version number of testtool too

File: src/healthcheck_dummy.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                                 void *arg) {
+  Healthcheck_dummy *hc = (Healthcheck_dummy *)arg;
+  string message = "dummy result";
+  HealthcheckResult result = HealthcheckResult::HC_PANIC;
+  hc->end_check(result, message);
+}
+
+int Healthcheck_dummy::schedule_healthcheck(struct timespec *now) {
+  // Peform general stuff for scheduled healthcheck.
+  if (Healthcheck::schedule_healthcheck(now) == false)
```
Comment (Woellchen) [inline src/healthcheck_dummy.cpp]: shortcut: `return Healthcheck::schedule_healthcheck(now)`

File: src/healthcheck_dummy.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    return false;
+
+  return true;
+}
+
+void Healthcheck_dummy::dummy_end_check(HealthcheckResult result,
+                                        string message) {
+  end_check(result, message);
+}
```
Comment (Woellchen) [inline src/healthcheck_dummy.cpp]: newline

File: src/healthcheck_http.h [hunk 4]
```diff
@@ -16,6 +16,8 @@
 
 #include "healthcheck.h"
 
+#define HC_UA "testtool"
```
Comment (Woellchen) [inline src/healthcheck_http.h:19]: how about to include testtool version number here? or somewhere else. but a typical user agent looks like `NAME/VERSION`

File: src/lb_node.cpp [hunk 5]
```diff
@@ -21,6 +21,20 @@ using namespace std;
 extern bool pf_action;
 extern int verbose;
 
+LbNodeAdminState admin_state_from_config(string s) {
+  if (s == "online")
```
Comment (Woellchen) [inline src/lb_node.cpp:27]: since `STATE_ENABLED` is the default, you can remove ifs for `online` and `deploy_online`

Comment (Woellchen) [reply src/lb_node.cpp:27]: or why not use a `switch` and group states together?

File: src/lb_node.cpp [hunk 6]
```diff
@@ -49,38 +64,31 @@ LbNode::LbNode(string name, const nlohmann::json &config,
...
   else
-    this->state = STATE_DOWN;
+    this->state = LbNodeState::STATE_DOWN;
+
   this->checked = false;
   this->min_nodes_kept = false;
   this->max_nodes_kept = false;
 
-  this->parent_lbpool->nodes.push_back(this);
+  this->parent_lbpool->nodes.insert(this);
```
Comment (Woellchen) [inline src/lb_node.cpp:79]: any special reason to use `insert` now?

File: src/lb_node.cpp [hunk 7]
```diff
@@ -112,38 +120,83 @@ void LbNode::finalize_healthchecks() {
...
+    state = LbNodeState::STATE_DOWN;
+    log(MessageType::MSG_STATE_DOWN, this,
         fmt::sprintf("message: %d of %d checks failed",
                      num_healthchecks - ok_healthchecks, num_healthchecks));
-  } else if (state == STATE_DOWN && new_state == STATE_UP) {
+  } else if (state == LbNodeState::STATE_DOWN &&
+             new_state == LbNodeState::STATE_UP) {
+
+    /*
+      LbNodeState new_state = LbNodeState::STATE_DOWN;
```
Comment (Woellchen) [inline src/lb_node.cpp]: any reason why we keep dead code here? i would rather keep the code base clean and instead use git history if you need something from the past

File: src/lb_pool.cpp [hunk 8]
```diff
@@ -289,8 +321,18 @@ void LbPool::update_pfctl(void) {
...
+set<string> LbPool::get_up_nodes_names() {
+  set<string> ret;
+  for (LbNode *node : up_nodes) {
+    ret.insert(node->name);
+  }
+  return ret;
+}
+
+set<LbNode *> LbPool::get_up_nodes() {
+  return up_nodes; // Hopefully it returns a copy
```
Comment (Woellchen) [inline src/lb_pool.cpp]: i cant accept the word `hopefully` here :D please make sure it is copied when you expect a copy

**Comment by Woellchen** [comment]
only minor points from my side. i like!

**Review by Woellchen — APPROVED** [review]

File: src/healthcheck.cpp [hunk 9]
```diff
@@ -93,13 +94,13 @@ Healthcheck::Healthcheck(const nlohmann::json &config,
   // Initialize healthchecks state basing on state of parent node.
   // Proper initial state for the healthcheck quarantees no
```
Comment (Felixoid) [inline src/healthcheck.cpp]:
```suggestion
  // Proper initial state for the healthcheck guarantees no
```

File: src/config.cpp [hunk 10]
```diff
@@ -29,5 +29,5 @@ template bool safe_get(const nlohmann::json &, const char *, bool);
 template int safe_get(const nlohmann::json &, const char *, int);
 template std::string safe_get(const nlohmann::json &, const char *,
                               std::string);
-template std::vector<std::string> safe_get(const nlohmann::json &, const char *,
-                                           std::vector<std::string>);
\ No newline at end of file
+template std::vector<int> safe_get(const nlohmann::json &, const char *,
+                                   std::vector<int>);
```
Comment (Felixoid) [inline src/config.cpp]: Sorry, but `noeol` could cause problems. I suggest add it here

File: src/healthcheck_http.cpp [hunk 11]
```diff
@@ -52,15 +52,26 @@ Healthcheck_http::Healthcheck_http(const nlohmann::json &config,
...
     this->ok_codes.push_back(fmt::sprintf("%d", ok_code));
   }
   if (this->ok_codes.size() == 0) {
-    this->ok_codes.push_back("200");
+    this->ok_codes.push_back(fmt::sprintf("%d", def_ok_codes[0]));
   }
+
+  vector<int> def_drain_codes{};
+  for (const int &drain_code :
+       safe_get<std::vector<int>>(config, "hc_drain_codes", def_drain_codes)) {
```
Comment (Felixoid) [inline src/healthcheck_http.cpp:71]: Good, good, good!

File: src/healthcheck_http.cpp [hunk 12]
```diff
@@ -256,16 +267,21 @@ void Healthcheck_http::event_callback(struct bufferevent *bev, short events,
 
   message = fmt::sprintf("HTTP code %s", statusline);
 
+  for (auto drain_code : hc->drain_codes)
```
Comment (Felixoid) [inline src/healthcheck_http.cpp:272]: Not sure, why not `int` here? and for ok_codes as well?

File: src/lb_pool.cpp [hunk 13]
```diff
@@ -178,62 +177,95 @@ void LbPool::pool_logic(LbNode *last_node) {
...
-                             node->name));
-            wanted_nodes.insert(node);
-          }
+
+        // Not enough nodes? Add those which wer force-kept previously.
+        for (LbNode *fuc : force_up_candidates) {
+          if (wanted_nodes.size() >= min_nodes)
+            break; // Enough nodes already wanted.
+          if (!fuc->min_nodes_kept)
+            continue; // Don't add nodes which were
```
Comment (Felixoid) [inline src/lb_pool.cpp]: Unfinished sentence?

File: src/lb_node.cpp (hunk 5 above)
Comment (TuxPowered42) [reply src/lb_node.cpp:27]: There's no swith/case for strings, it can only work for simple types.

File: src/lb_node.cpp (hunk 6 above)
Comment (TuxPowered42) [reply src/lb_node.cpp:79]: The variable was converted from vector to a set, because there must be no assumption about ordering.

**Review by Felixoid — COMMENTED** [review]
There are few questions and few suggestions

File: src/healthcheck_http.cpp (hunk 12 above)
Comment (TuxPowered42) [reply src/healthcheck_http.cpp:272]: Usage of strings is documented in line 58 when they are loaded from json.

File: src/config.cpp (hunk 10 above)
Comment (TuxPowered42) [reply src/config.cpp]: Fixed.

File: src/healthcheck.cpp (hunk 9 above)
Comment (TuxPowered42) [reply src/healthcheck.cpp]: Fixed.

File: src/lb_pool.cpp (hunk 13 above)
Comment (TuxPowered42) [reply src/lb_pool.cpp]: Fixed.

File: src/healthcheck_dummy.cpp (hunk 2 above)
Comment (TuxPowered42) [reply src/healthcheck_dummy.cpp]: Fixed.

File: src/lb_node.cpp (hunk 7 above)
Comment (TuxPowered42) [reply src/lb_node.cpp]: It was a mistake.

File: src/healthcheck_dummy.cpp (hunk 3 above)
Comment (TuxPowered42) [reply src/healthcheck_dummy.cpp]: Fixed.

File: src/lb_pool.cpp (hunk 8 above)
Comment (TuxPowered42) [reply src/lb_pool.cpp]: It's not a set, not a pointer to one, so it will be copied.

File: src/lb_node.cpp (hunk 6 above)
Comment (Woellchen) [reply src/lb_node.cpp:79]: i missed that detail. nevermind

File: src/lb_node.cpp (hunk 5 above)
Comment (Woellchen) [reply src/lb_node.cpp:27]: sorry haven't done c/++ for too long. we can still remove the two redundant ifs then

**Review by Felixoid — APPROVED** [review]

**Pull request merged**